cmd/scorer/main.go

```diff
@@ -0,0 +1,231 @@
+}
+
+func makeOutHeader(header []string, resultColumn string) ([]string, error) {
+	for _, h := range header {
+		if h == resultColumn {
+			return nil, fmt.Errorf("header already contains field %s", resultColumn)
+		}
+	}
+	return append(header, resultColumn), nil
+}
+
+func makeRecord(header []string, row []string) map[string]float64 {
+	record := make(map[string]float64)
+	for i, k := range header {
+		raw := row[i]
+		v, err := strconv.ParseFloat(raw, 64)
+		if err != nil {
+			// Failed to parse raw into a float, ignore the field
+			continue
+		}
+		record[k] = v
+	}
+	return record
+}
+
+func main() {
+	flag.Parse()
+
+	logger := log.New()
+	logger.SetLevel(logFlag.Level())
+
+	if flag.NArg() != 2 {
+		logger.Error("Must have an input file and an output file specified")
+		os.Exit(2)
+	}
+	inFilename := flag.Args()[0]
+	outFilename := flag.Args()[1]
+
+	// Open the in-file for reading
+	var r *csv.Reader
+	if inFilename == "-" {
+		logger.Info("Reading from stdin")
+		r = csv.NewReader(os.Stdin)
+	} else {
+		logger.WithFields(log.Fields{
+			"filename": inFilename,
+		}).Debug("Reading from file")
+		f, err := os.Open(inFilename)
+		if err != nil {
+			logger.WithFields(log.Fields{
+				"error":    err,
+				"filename": inFilename,
+			}).Error("Failed to open input file")
+			os.Exit(2)
+		}
+		defer f.Close()
+		r = csv.NewReader(f)
+	}
+
+	// Open the out-file for writing
+	f, err := outfile.Open(outFilename)
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error":    err,
+			"filename": outFilename,
+		}).Error("Failed to open file for output")
+		os.Exit(2)
+	}
+	defer f.Close()
+	w := csv.NewWriter(f)
+	defer w.Flush()
+
+	// Prepare the algorithm from the config file
+	if *configFlag == "" {
+		logger.Error("Must have a config file set")
+		os.Exit(2)
+	}
+
+	f, err = os.Open(*configFlag)
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error":    err,
+			"filename": configFlag,
+		}).Error("Failed to open config file")
+		os.Exit(2)
+	}
+	c, err := LoadConfig(f)
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error":    err,
+			"filename": configFlag,
+		}).Error("Failed to parse config file")
+		os.Exit(2)
+	}
+	a, err := c.Algorithm()
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error":     err,
+			"algorithm": c.Name,
+		}).Error("Failed to get the algorithm")
+		os.Exit(2)
+	}
+
+	inHeader, err := r.Read()
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error": err,
+		}).Error("Failed to read CSV header row")
+		os.Exit(2)
+	}
+
+	// Generate and output the CSV header row
+	outHeader, err := makeOutHeader(inHeader, generateColumnName())
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error": err,
+		}).Error("Failed to generate output header row")
+		os.Exit(2)
+	}
+	if err := w.Write(outHeader); err != nil {
+		logger.WithFields(log.Fields{
+			"error": err,
+		}).Error("Failed to write CSV header row")
+		os.Exit(2)
+	}
+
+	var pq PriorityQueue
```

Comment:
I agree in principal that this would be better outside of this tool, but the ergonomics of `sort` isn't that great.

`sort` doesn't handle the header row nicely, and the column is hard to specify.

I'm going to leave this in for now. I have in mind preparing a collection of tools for analysing the csv data (e.g. stats like CDF, median, min, max, correlation, and the differences in ranking between two sets of scores) and this could exist alongside it.

Reply:
ack!